### ISSUE TITLE
Is there AllReduce in data parallelism?

### ISSUE THREAD

**Issue by Allen-Czyysx** [msr-fiddle/pipedream#65]
I am wondering where is the logic of allreduce in the training of PipeDream.

It seems PipeDream uses `torch.nn.parallel.DistributedDataParallel` in `runtime.py`, but also uses `torch.autograd.grad`, which is against https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html.

I would appreciate it if there is any hint to help me figure out where the communication happens in the data parallelism of PipeDream.

**Comment by deepakn94** [comment]
Yeah, we use `torch.nn.parallel.DistributedDataParallel` to implement all-reduces.

We don't use `torch.autograd.grad`, we use `torch.autograd.backward`: https://github.com/msr-fiddle/pipedream/blob/pipedream/runtime/runtime.py#L606.

**Comment by Allen-Czyysx** [comment]
Thanks for your reply!

But I am still confused. According to https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html, `DistributedDataParallel` should be used along with `torch.distributed.autograd.backward`.

In my understanding, both `DistributedDataParallel` and `torch.distributed.autograd.backward` contribute to allreduce.

**Comment by deepakn94** [comment]
I don't understand your question.

The `DistributedDataParallel` wrapper inserts appropriate backward hooks so that an all-reduce is initiated any time `torch.distributed.autograd.backward()` is called (or alternatively, when `loss.backward()` is called).

**Comment by Allen-Czyysx** [comment]
My confusion here is what the difference between `DistributedDataParallel` with `torch.distributed.autograd.backward` and `DistributedDataParallel` with `torch.autograd.backward`.

**Comment by deepakn94** [comment]
I see. The way I understand it, `torch.distributed.autograd.backward` is necessary if you want to use PyTorch's RPC interface (with `send` and `recv` nodes in the implicit operator graph). This is not necessary for regular data parallelism. We do not use `torch.distributed.autograd.backward` anywhere in this codebase.

**Comment by Allen-Czyysx** [comment]
Thanks. That helps a lot : )
